File: android/app/src/main/java/com/lumi/dm/engine/DownloadEngine.java
package com.lumi.dm.engine;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class DownloadEngine {

    public interface Callback {
        void onUpdate(DownloadJob job);
    }

    public interface Listener {
        void onJobsChanged();
    }

    private static DownloadEngine instance;

    private final Map<String, DownloadJob>         jobs      = Collections.synchronizedMap(new LinkedHashMap<>());
    private final Map<String, HttpDownloadTask>     tasks     = Collections.synchronizedMap(new LinkedHashMap<>());
    private       ExecutorService                   executor  = Executors.newFixedThreadPool(4);
    private       int                               maxConc          = 3;
    private       int                               defaultConns     = 16;
    private final List<Listener>                    listeners = new ArrayList<>();
    private final Handler                           mainHandler = new Handler(Looper.getMainLooper());
    private       Context                           appContext;
    private       String                            defaultDir;

    private boolean initialized = false;

    private DownloadEngine() {}

    public static synchronized DownloadEngine get() {
        if (instance == null) instance = new DownloadEngine();
        return instance;
    }

    public synchronized void init(Context ctx) {
        if (initialized) return;
        initialized = true;
        appContext = ctx.getApplicationContext();
        defaultDir = Environment.getExternalStoragePublicDirectory(
                Environment.DIRECTORY_DOWNLOADS).getAbsolutePath();

        SharedPreferences prefs = ctx.getSharedPreferences("LUMIDM_prefs", Context.MODE_PRIVATE);
        defaultDir   = prefs.getString("download_dir", defaultDir);
        maxConc      = prefs.getInt("max_concurrent", 3);
        defaultConns = prefs.getInt("default_connections", 16);
        recreatePool();
        loadState();
    }

    private void recreatePool() {
        if (!executor.isShutdown()) executor.shutdown();
        executor = Executors.newFixedThreadPool(Math.max(1, maxConc));
    }

    // ── Listeners ─────────────────────────────────────────────────────────────

    public void addListener(Listener l)    { listeners.add(l); }
    public void removeListener(Listener l) { listeners.remove(l); }

    private void notifyListeners() {
        mainHandler.post(() -> {
            for (Listener l : new ArrayList<>(listeners)) l.onJobsChanged();
        });
    }

    // ── Job accessors ─────────────────────────────────────────────────────────

    public List<DownloadJob> getJobs() {
        List<DownloadJob> list = new ArrayList<>(jobs.values());
        Collections.reverse(list);
        return list;
    }

    public DownloadJob getJob(String id) { return jobs.get(id); }

    // ── Add downloads ──────────────────────────────────────────────────────────

    public DownloadJob addHttp(String url, String filename, String dir) {
        if (dir == null || dir.isEmpty()) dir = defaultDir;
        if (filename == null || filename.isEmpty()) filename = guessFilename(url);
        DownloadJob job = new DownloadJob(url, filename, dir, DownloadJob.Type.HTTP);
        job.connections = defaultConns;
        enqueue(job);
        return job;
    }

    public DownloadJob addVideo(String url, String filename, String dir) {
        if (dir == null || dir.isEmpty()) dir = defaultDir;
        if (filename == null || filename.isEmpty()) filename = "video_" + System.currentTimeMillis() + ".mp4";
        DownloadJob job = new DownloadJob(url, filename, dir, DownloadJob.Type.VIDEO);
        enqueue(job);
        return job;
    }

    private void enqueue(DownloadJob job) {
        jobs.put(job.id, job);
        saveState();
        submitTask(job);
        notifyListeners();
    }

    private void submitTask(DownloadJob job) {
        if (job.type == DownloadJob.Type.HTTP || job.type == DownloadJob.Type.VIDEO) {
            HttpDownloadTask task = new HttpDownloadTask(job, updatedJob -> {
                saveState();
                notifyListeners();
            });
            tasks.put(job.id, task);
            executor.submit(task);
        }
    }

    // ── Controls ───────────────────────────────────────────────────────────────

    public void pause(String id) {
        HttpDownloadTask t = tasks.get(id);
        if (t != null) t.pause();
    }

    public void resume(String id) {
        DownloadJob job = jobs.get(id);
        if (job == null) return;
        HttpDownloadTask t = tasks.get(id);
        if (t != null) {
            t.resume();
        } else if (job.status == DownloadJob.Status.PAUSED || job.status == DownloadJob.Status.FAILED) {
            job.status = DownloadJob.Status.QUEUED;
            submitTask(job);
            notifyListeners();
        }
    }

    public void cancel(String id) {
        HttpDownloadTask t = tasks.get(id);
        if (t != null) t.cancel();
        DownloadJob job = jobs.get(id);
        if (job != null && !job.isTerminal()) {
            job.status = DownloadJob.Status.CANCELLED;
            saveState();
            notifyListeners();
        }
    }

    public void delete(String id) {
        cancel(id);
        jobs.remove(id);
        tasks.remove(id);
        saveState();
        notifyListeners();
    }

    public void pauseAll() {
        for (DownloadJob j : jobs.values()) if (j.status == DownloadJob.Status.RUNNING) pause(j.id);
    }

    public void resumeAll() {
        for (DownloadJob j : jobs.values()) if (j.status == DownloadJob.Status.PAUSED) resume(j.id);
    }

    public void clearCompleted() {
        List<String> toRemove = new ArrayList<>();
        for (DownloadJob j : jobs.values()) if (j.isTerminal()) toRemove.add(j.id);
        for (String id : toRemove) { jobs.remove(id); tasks.remove(id); }
        saveState();
        notifyListeners();
    }

    // ── Settings ───────────────────────────────────────────────────────────────

    public String getDefaultDir() { return defaultDir; }

    public void setDefaultDir(String dir) {
        defaultDir = dir;
        savePrefs();
    }

    public void setMaxConcurrent(int n) {
        maxConc = Math.max(1, Math.min(16, n));
        recreatePool();
        savePrefs();
    }

    public int getMaxConcurrent() { return maxConc; }

    public void setDefaultConnections(int n) {
        defaultConns = Math.max(1, Math.min(64, n));
        savePrefs();
    }

    public int getDefaultConnections() { return defaultConns; }

    private void savePrefs() {
        if (appContext == null) return;
        appContext.getSharedPreferences("LUMIDM_prefs", Context.MODE_PRIVATE)
            .edit()
            .putString("download_dir", defaultDir)
            .putInt("max_concurrent", maxConc)
            .putInt("default_connections", defaultConns)
            .apply();
    }

    // ── Persistence ────────────────────────────────────────────────────────────

    private File stateFile() {
        return new File(appContext.getFilesDir(), "downloads.json");
    }

    private void saveState() {
        if (appContext == null) return;
        try {
            JSONArray arr = new JSONArray();
            for (DownloadJob j : jobs.values()) {
                JSONObject o = new JSONObject();
                o.put("id",               j.id);
                o.put("url",              j.url);
                o.put("filename",         j.filename);
                o.put("targetDir",        j.targetDir);
                o.put("type",             j.type.name());
                o.put("status",           j.isTerminal() ? j.status.name() : DownloadJob.Status.PAUSED.name());
                o.put("totalBytes",       j.totalBytes);
                o.put("downloadedBytes",  j.downloadedBytes);
                o.put("connections",      j.connections);
                o.put("error",            j.error != null ? j.error : "");
                o.put("createdAt",        j.createdAt);
                arr.put(o);
            }
            try (FileWriter fw = new FileWriter(stateFile())) { fw.write(arr.toString()); }
        } catch (Exception ignored) {}
    }

    private void loadState() {
        File f = stateFile();
        if (!f.exists()) return;
        try {
            StringBuilder sb = new StringBuilder();
            try (BufferedReader br = new BufferedReader(new FileReader(f))) {
                String line;
                while ((line = br.readLine()) != null) sb.append(line);
            }
            JSONArray arr = new JSONArray(sb.toString());
            for (int i = 0; i < arr.length(); i++) {
                JSONObject o = arr.getJSONObject(i);
                DownloadJob j = new DownloadJob(
                    o.getString("url"),
                    o.getString("filename"),
                    o.getString("targetDir"),
                    DownloadJob.Type.valueOf(o.optString("type", "HTTP"))
                );
                // Reflect the saved id
                try {
                    java.lang.reflect.Field fid = DownloadJob.class.getDeclaredField("id");
                    fid.setAccessible(true);
                    fid.set(j, o.getString("id"));
                } catch (Exception ignored) {}

                j.status          = DownloadJob.Status.valueOf(o.optString("status", "PAUSED"));
                j.totalBytes      = o.optLong("totalBytes", 0);
                j.downloadedBytes = o.optLong("downloadedBytes", 0);
                j.connections     = o.optInt("connections", defaultConns);
                j.error           = o.optString("error", null);
                j.createdAt       = o.optLong("createdAt", System.currentTimeMillis());
                jobs.put(j.id, j);
            }
        } catch (Exception ignored) {}
    }

    // ── Helpers ────────────────────────────────────────────────────────────────

    private String guessFilename(String url) {
        try {
            String path = new java.net.URL(url).getPath();
            String name = path.substring(path.lastIndexOf('/') + 1);
            if (!name.isEmpty()) return name.split("\\?")[0];
        } catch (Exception ignored) {}
        return "download_" + System.currentTimeMillis();
    }

    public long totalSpeed() {
        long s = 0;
        for (DownloadJob j : jobs.values()) if (j.status == DownloadJob.Status.RUNNING) s += j.speedBps;
        return s;
    }
}

File: ios/LumiDM/DownloadEngine.swift
import Foundation

final class DownloadEngine {

    static let shared = DownloadEngine()

    // Broadcast when jobs change — observers register on main thread
    static let didUpdateNotification = Notification.Name("DownloadEngineDidUpdate")

    private(set) var jobs: [DownloadJob] = []
    private var tasks:     [String: HttpDownloadTask] = [:]

    var maxConcurrent: Int = 3 { didSet { savePrefs() } }
    var defaultDir: String    { didSet { savePrefs() } }

    private init() {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        defaultDir = docs.appendingPathComponent("Downloads").path
        try? FileManager.default.createDirectory(
            atPath: defaultDir, withIntermediateDirectories: true)
        loadPrefs()
        loadState()
    }

    // MARK: - Add

    @discardableResult
    func addHttp(url: String, filename: String? = nil, dir: String? = nil) -> DownloadJob {
        let name = filename ?? guessFilename(from: url)
        let job  = DownloadJob(url: url, filename: name,
                               targetDir: dir ?? defaultDir, type: .http)
        enqueue(job)
        return job
    }

    @discardableResult
    func addVideo(url: String, filename: String? = nil, dir: String? = nil) -> DownloadJob {
        let name = filename ?? "video_\(Int(Date().timeIntervalSince1970)).mp4"
        let job  = DownloadJob(url: url, filename: name,
                               targetDir: dir ?? defaultDir, type: .video)
        enqueue(job)
        return job
    }

    private func enqueue(_ job: DownloadJob) {
        jobs.insert(job, at: 0)
        saveState()
        startNextIfNeeded()
        broadcast()
    }

    // MARK: - Controls

    func pause(_ id: String) {
        tasks[id]?.pause()
    }

    func resume(_ id: String) {
        guard let job = jobs.first(where: { $0.id == id }),
              job.status == .paused || job.status == .failed else { return }
        job.status = .queued
        startNextIfNeeded()
        broadcast()
    }

    func cancel(_ id: String) {
        tasks[id]?.cancel()
        tasks.removeValue(forKey: id)
        if let job = jobs.first(where: { $0.id == id }), !job.isTerminal {
            job.status = .cancelled
            saveState()
            broadcast()
        }
    }

    func delete(_ id: String) {
        cancel(id)
        jobs.removeAll { $0.id == id }
        saveState()
        broadcast()
    }

    func pauseAll()      { jobs.filter { $0.status == .running  }.forEach { pause($0.id) } }
    func resumeAll()     { jobs.filter { $0.status == .paused   }.forEach { resume($0.id) } }
    func clearCompleted(){ jobs.filter { $0.isTerminal          }.forEach { delete($0.id) } }

    var totalSpeed: Int64 {
        jobs.filter { $0.status == .running }.reduce(0) { $0 + $1.speedBps }
    }

    // MARK: - Internal queue

    private func startNextIfNeeded() {
        let running = jobs.filter { $0.status == .running }.count
        guard running < maxConcurrent else { return }
        guard let next = jobs.first(where: { $0.status == .queued }) else { return }
        submit(next)
    }

    private func submit(_ job: DownloadJob) {
        let task = HttpDownloadTask(job: job) { [weak self] updated in
            self?.saveState()
            self?.broadcast()
            // Clean up task ref when done or paused (paused creates a fresh task on resume)
            if updated.isTerminal || updated.status == .paused {
                self?.tasks.removeValue(forKey: updated.id)
            }
            if updated.isTerminal {
                self?.startNextIfNeeded()
                if updated.status == .completed {
                    NotificationHelper.shared.sendCompletionIfAllDone(
                        jobs: self?.jobs ?? [])
                }
            }
        }
        tasks[job.id] = task
        task.start()
    }

    // MARK: - Broadcast

    private func broadcast() {
        DispatchQueue.main.async {
            NotificationCenter.default.post(
                name: DownloadEngine.didUpdateNotification, object: self)
        }
    }

    // MARK: - Persistence

    private var stateURL: URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("downloads.json")
    }

    private var prefsURL: URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("LUMIDM_prefs.json")
    }

    func saveState() {
        // Encode newest-first; on load we reverse
        if let data = try? JSONEncoder().encode(jobs) {
            try? data.write(to: stateURL)
        }
    }

    private func loadState() {
        guard let data  = try? Data(contentsOf: stateURL),
              let saved = try? JSONDecoder().decode([DownloadJob].self, from: data)
        else { return }
        // Mark any in-progress jobs as paused (app was killed mid-download)
        saved.forEach { if $0.status == .running { $0.status = .paused } }
        jobs = saved
    }

    func savePrefs() {
        let d: [String: Any] = ["defaultDir": defaultDir, "maxConcurrent": maxConcurrent]
        if let data = try? JSONSerialization.data(withJSONObject: d) {
            try? data.write(to: prefsURL)
        }
    }

    private func loadPrefs() {
        guard let data  = try? Data(contentsOf: prefsURL),
              let prefs = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return }
        if let dir = prefs["defaultDir"]    as? String { defaultDir    = dir }
        if let max = prefs["maxConcurrent"] as? Int    { maxConcurrent = max }
    }

    // MARK: - Helpers

    private func guessFilename(from urlStr: String) -> String {
        let name = URL(string: urlStr)?.lastPathComponent ?? ""
        if !name.isEmpty && name != "/" {
            return name.components(separatedBy: "?")[0]
        }
        return "download_\(Int(Date().timeIntervalSince1970))"
    }
}
